raft-tech/TANF-app
# Issue 1165: Resolve Cache Control header error by ZAP

### Summary of Changes
Closes #1165 

Adds a [Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) header to both the backend and frontend applications in order to remediate an alert found by the OWASP ZAP scanner.

For the backend, a custom Django middleware has been added which will add the appropriate headers needed to disable caching of protected resources to all HTTP responses handled by the backend application. On the frontend side, the header was manually added to the nginx config that gets used for the Cloud.gov buildpack.

### How to Test
This branch must be tested against applications deployed in Cloud.gov.
1. Deploy this PR to your designated testing application in the Cloud.gov dev space.
2. Modify your local zap-scanner.sh script on line 14 such that it now directs scans towards the URL of the site you deployed to. ie:
```
APP_URL="https://tdp-$TARGET-raft.app.cloud.gov"
```
3. Execute the zap-scanner script against your target:
```
./scripts/zap-scanner.sh backend nightly
./scripts/zap-scanner.sh frontend nightly
```
4. Open up the generated report files in `tdrs-backend/reports` and `tdrs-frontend/reports`
5. Confirm that the `Invalid or no Cache-Control header` alert is no longer raised on the report.

Example reports:
![Screenshot from 2021-11-16 13-41-09](https://user-images.githubusercontent.com/22626085/142046946-ee9fc4f4-06f1-40b8-976a-ae6e7232e324.png)
![Screenshot from 2021-11-16 13-41-02](https://user-images.githubusercontent.com/22626085/142046948-ed723102-e091-4197-82fd-2ff3b941eae0.png)

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

  - [x] Zap report no longer features this error.

As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*


### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*
    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/Technical-Documentation/how-government-will-test-a11y.md)

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful deployment by assigning a 'Deploy with CircleCI - {env}' label*


### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*


### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*


### Deliverable 8: Context

> * Performance Standard(s): Code must be understandable and contextualized for the reviewers possess the knowledge and background necessary for analysis and constructive criticism to take place.

> * Acceptable Quality Level: Code submitted in the pull request has context.*
